db/migrate/20210609081621_add_score_to_profiles.rb

```diff
@@ -0,0 +1,10 @@
+class AddScoreToProfiles < ActiveRecord::Migration[5.2]
+  def up
+    add_column :profiles, :score, :decimal
+    Profile.canonical(false).where(score: nil).joins(:test_results).distinct.find_each(&:calculate_score!)
```

Comment:
This will likely be very heavy on the database. I have 9 non-canonical profiles (only one host assigned) and it does 16 queries:
```
[2] pry(main)> Profile.canonical(false).joins(:test_results).distinct.find_each(&:calculate_score!)
  Profile Load (2.0ms)  SELECT  DISTINCT "profiles".* FROM "profiles" INNER JOIN "test_results" ON "test_results"."profile_id" = "profiles"."id" WHERE "profiles"."parent_profile_id" IS NOT NULL ORDER BY "profiles"."id" ASC LIMIT $1  [["LIMIT", 1000]]
   (0.8ms)  SELECT AVG("test_results"."score") FROM "test_results" JOIN (SELECT "test_results"."profile_id", "test_results"."host_id", MAX(end_time) as end_time FROM "test_results" WHERE "test_results"."profile_id" = 'df0d257f-169c-473e-a4cb-82b854c3f342' GROUP BY "test_results"."profile_id", "test_results"."host_id") as tr on test_results.profile_id = tr.profile_id AND test_results.host_id = tr.host_id AND test_results.end_time = tr.end_time WHERE "test_results"."profile_id" = $1  [["profile_id", "df0d257f-169c-473e-a4cb-82b854c3f342"]]
   (0.2ms)  BEGIN
  Xccdf::Benchmark Load (0.4ms)  SELECT  "benchmarks".* FROM "benchmarks" WHERE "benchmarks"."id" = $1 LIMIT $2  [["id", "e7f39c61-40aa-4fe7-b8db-70b6764ca3e6"], ["LIMIT", 1]]
  Profile Exists (0.6ms)  SELECT  1 AS one FROM "profiles" WHERE "profiles"."ref_id" = $1 AND "profiles"."id" != $2 AND "profiles"."account_id" = $3 AND "profiles"."benchmark_id" = $4 AND "profiles"."os_minor_version" = $5 AND "profiles"."policy_id" = $6 LIMIT $7  [["ref_id", "xccdf_org.ssgproject.content_profile_C2S"], ["id", "df0d257f-169c-473e-a4cb-82b854c3f342"], ["account_id", "7f544c34-8d4d-4edb-8b77-09bb6ce0f8f7"], ["benchmark_id", "e7f39c61-40aa-4fe7-b8db-70b6764ca3e6"], ["os_minor_version", ""], ["policy_id", "e29c54de-6e9e-48ec-8bca-0e391b5e3b10"], ["LIMIT", 1]]
  Host Exists (12.9ms)  SELECT  1 AS one FROM "inventory"."hosts" INNER JOIN "test_results" ON "inventory"."hosts"."id" = "test_results"."host_id" WHERE "test_results"."profile_id" = $1 LIMIT $2  [["profile_id", "df0d257f-169c-473e-a4cb-82b854c3f342"], ["LIMIT", 1]]
  Account Load (0.4ms)  SELECT  "accounts".* FROM "accounts" WHERE "accounts"."id" = $1 LIMIT $2  [["id", "7f544c34-8d4d-4edb-8b77-09bb6ce0f8f7"], ["LIMIT", 1]]
  Policy Load (0.3ms)  SELECT  "policies".* FROM "policies" WHERE "policies"."id" = $1 LIMIT $2  [["id", "e29c54de-6e9e-48ec-8bca-0e391b5e3b10"], ["LIMIT", 1]]
   (0.3ms)  COMMIT
   (0.6ms)  SELECT AVG("test_results"."score") FROM "test_results" JOIN (SELECT "test_results"."profile_id", "test_results"."host_id", MAX(end_time) as end_time FROM "test_results" WHERE "test_results"."profile_id" = 'ed19d644-d57e-47cc-9e6a-9f46729f2482' GROUP BY "test_results"."profile_id", "test_results"."host_id") as tr on test_results.profile_id = tr.profile_id AND test_results.host_id = tr.host_id AND test_results.end_time = tr.end_time WHERE "test_results"."profile_id" = $1  [["profile_id", "ed19d644-d57e-47cc-9e6a-9f46729f2482"]]
   (0.2ms)  BEGIN
  Xccdf::Benchmark Load (0.2ms)  SELECT  "benchmarks".* FROM "benchmarks" WHERE "benchmarks"."id" = $1 LIMIT $2  [["id", "e7f39c61-40aa-4fe7-b8db-70b6764ca3e6"], ["LIMIT", 1]]
  Profile Exists (0.4ms)  SELECT  1 AS one FROM "profiles" WHERE "profiles"."ref_id" = $1 AND "profiles"."id" != $2 AND "profiles"."account_id" = $3 AND "profiles"."benchmark_id" = $4 AND "profiles"."os_minor_version" = $5 AND "profiles"."policy_id" = $6 LIMIT $7  [["ref_id", "xccdf_org.ssgproject.content_profile_hipaa"], ["id", "ed19d644-d57e-47cc-9e6a-9f46729f2482"], ["account_id", "7f544c34-8d4d-4edb-8b77-09bb6ce0f8f7"], ["benchmark_id", "e7f39c61-40aa-4fe7-b8db-70b6764ca3e6"], ["os_minor_version", ""], ["policy_id", "ea50e400-a460-4cce-b505-ca1fa70b599c"], ["LIMIT", 1]]
  Host Exists (11.0ms)  SELECT  1 AS one FROM "inventory"."hosts" INNER JOIN "test_results" ON "inventory"."hosts"."id" = "test_results"."host_id" WHERE "test_results"."profile_id" = $1 LIMIT $2  [["profile_id", "ed19d644-d57e-47cc-9e6a-9f46729f2482"], ["LIMIT", 1]]
  Account Load (0.5ms)  SELECT  "accounts".* FROM "accounts" WHERE "accounts"."id" = $1 LIMIT $2  [["id", "7f544c34-8d4d-4edb-8b77-09bb6ce0f8f7"], ["LIMIT", 1]]
  Policy Load (0.3ms)  SELECT  "policies".* FROM "policies" WHERE "policies"."id" = $1 LIMIT $2  [["id", "ea50e400-a460-4cce-b505-ca1fa70b599c"], ["LIMIT", 1]]
   (0.2ms)  COMMIT
   (0.6ms)  SELECT AVG("test_results"."score") FROM "test_results" JOIN (SELECT "test_results"."profile_id", "test_results"."host_id", MAX(end_time) as end_time FROM "test_results" WHERE "test_results"."profile_id" = 'faf7e0b2-f02d-4f6f-b530-473bb1aef85f' GROUP BY "test_results"."profile_id", "test_results"."host_id") as tr on test_results.profile_id = tr.profile_id AND test_results.host_id = tr.host_id AND test_results.end_time = tr.end_time WHERE "test_results"."profile_id" = $1  [["profile_id", "faf7e0b2-f02d-4f6f-b530-473bb1aef85f"]]
   (0.2ms)  BEGIN
  Xccdf::Benchmark Load (0.3ms)  SELECT  "benchmarks".* FROM "benchmarks" WHERE "benchmarks"."id" = $1 LIMIT $2  [["id", "e7f39c61-40aa-4fe7-b8db-70b6764ca3e6"], ["LIMIT", 1]]
  Profile Exists (0.4ms)  SELECT  1 AS one FROM "profiles" WHERE "profiles"."ref_id" = $1 AND "profiles"."id" != $2 AND "profiles"."account_id" = $3 AND "profiles"."benchmark_id" = $4 AND "profiles"."os_minor_version" = $5 AND "profiles"."policy_id" = $6 LIMIT $7  [["ref_id", "xccdf_org.ssgproject.content_profile_cis"], ["id", "faf7e0b2-f02d-4f6f-b530-473bb1aef85f"], ["account_id", "7f544c34-8d4d-4edb-8b77-09bb6ce0f8f7"], ["benchmark_id", "e7f39c61-40aa-4fe7-b8db-70b6764ca3e6"], ["os_minor_version", ""], ["policy_id", "f43f2b47-5afb-4b3d-ab81-3a746863bb76"], ["LIMIT", 1]]
  Host Exists (11.2ms)  SELECT  1 AS one FROM "inventory"."hosts" INNER JOIN "test_results" ON "inventory"."hosts"."id" = "test_results"."host_id" WHERE "test_results"."profile_id" = $1 LIMIT $2  [["profile_id", "faf7e0b2-f02d-4f6f-b530-473bb1aef85f"], ["LIMIT", 1]]
  Account Load (0.5ms)  SELECT  "accounts".* FROM "accounts" WHERE "accounts"."id" = $1 LIMIT $2  [["id", "7f544c34-8d4d-4edb-8b77-09bb6ce0f8f7"], ["LIMIT", 1]]
  Policy Load (0.3ms)  SELECT  "policies".* FROM "policies" WHERE "policies"."id" = $1 LIMIT $2  [["id", "f43f2b47-5afb-4b3d-ab81-3a746863bb76"], ["LIMIT", 1]]
   (0.1ms)  COMMIT
```

Reply:
This one runs only once, but I'm open for any kind of optimizations here.

Reply:
So the only truly efficient way to optimize this into two queries needs to use some raw sql :disappointed: which might change if we upgrade to Rails 6 as it supports `upsert`. It probably has to be wrapped around a transaction and needs to be prettified, but:
```ruby
scores = TestResult.latest.joins(:profile).group(:profile_id).average(:score)
values = scores.map { |k, v| "(#{ActiveRecord::Base.connection.quote(k)}, #{ActiveRecord::Base.connection.quote(v)})" }.join(', ')

query = "
    UPDATE profiles T
    SET score = uv.new_score
    FROM (VALUES #{values}) AS uv (id, new_score)
    WHERE T.id = uv.id::uuid"

ActiveRecord::Base.connection.execute(query)
```
@akofink what do you think?

Reply:
I'm OK with using raw SQL like this for a one-off migration if it's our only option. Let me investigate it a bit more - I haven't looked especially closely at this work. I will get back to you :)

Reply:
I tried benchmarking this original implementation, and it doesn't seem to be that problematic:

```
$ time podman exec -it compliance-backend_prometheus_1 bundle exec rake db:migrate
`Redis#exists(key)` will return an Integer by default in redis-rb 4.3. The option to explicitly disable this behaviour via `Redis.exists_returns_integer` will be removed in 5.0. You should use `exists?` instead.
2021-06-18T17:17:48.125Z 5300 TID-8go INFO: GitLab reliable fetch activated!
== 20210609081621 AddScoreToProfiles: migrating ===============================
-- add_column(:profiles, :score, :decimal)
   -> 0.0009s
== 20210609081621 AddScoreToProfiles: migrated (0.3996s) ======================

podman exec -it compliance-backend_prometheus_1 bundle exec rake db:migrate  0.11s user 0.14s system 10% cpu 2.365 total
```

Under 2.5 seconds for 6033 hosts with 5930 latest test results.